src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -960,32 +961,48 @@ private AnonymousContentAccessCertificate createAnonContentAccessCertificate(
     }
 
     private String createAnonPayloadAndSignature(Collection<ProductInfo> prodInfo) throws IOException {
-        List<ContentInfo> contentInfo = getContentInfo(prodInfo);
-        List<org.candlepin.model.Content> contents = convertContentInfoToContent(contentInfo);
-        Map<org.candlepin.model.Content, Boolean> activeContent = new HashMap<>();
-        contents.forEach(content -> activeContent.put(content, true));
+        Map<ContentInfo, Boolean> contentInfo = getContentInfo(prodInfo);
+        Map<org.candlepin.model.Content, Boolean> activeContent = convertContentInfoToContent(contentInfo);
         PromotedContent promotedContent = new PromotedContent(new AnonymousContentPrefix());
         byte[] data = createContentAccessDataPayload(null, activeContent, promotedContent);
 
         return createPayloadAndSignature(data);
     }
 
     /**
-     * Retrieves all of the {@link ContentInfo} from the provided {@link ProductInfo}
+     * Retrieves all of the {@link ContentInfo} and their enabled (true/false) status from the provided
+     * {@link ProductInfo}. In case more than one of the same content, but different enabled
+     * values are provided, the enabled=true value is kept.
      *
      * @param prodInfo
      *  the product info that contains content info
      *
-     * @return all of the content info from the provided product info
+     * @return a map that contains all the content info and their enabled/disabled status based on the
+     * provided product info
      */
-    private List<ContentInfo> getContentInfo(Collection<ProductInfo> prodInfo) {
-        return prodInfo.stream()
+    private Map<ContentInfo, Boolean> getContentInfo(Collection<ProductInfo> prodInfo) {
+        Collection<ProductContentInfo> productContentInfos = prodInfo.stream()
             .filter(Objects::nonNull)
             .map(ProductInfo::getProductContent)
             .flatMap(Collection::stream)
             .filter(Objects::nonNull)
-            .map(ProductContentInfo::getContent)
-            .toList();
+            .collect(Collectors.toList());
```

Comment:
We can still use a `.collect(Collectors.toMap(...))` here, just need to use the version that takes a merge function:

```
.collect(Collectors.toMap(ProductContentInfo::getContent, ProductContentInfo::isEnabled, (e, u) -> { e || u }))
```

This also avoids doubly iterating through the collection.

The other concern here is whether or not ContentInfo will even collide as a key in the way we want/expect for resolution. Since it's just an interface, we're hoping that each instance properly implements hashCode and equals.

We may be in a situation where our list always has duplicates in the case of multiple products with the same content. Though in the case of specifically SCA mode, we're mostly isolated by known properties of the upstream data (no content is shared).

Maybe this is fine for now, but I think in terms of our actual data model and long-term maintenance, this is a ticking time bomb.

Reply:
Nice, I didn't know about the merge function. Updated.

With regards to how collisions are handled, the adapter where this data is coming from today is using our ProductContentData/ContentData concrete implementations of the *Info interfaces, which do have proper hashcode/equals

Reply:
Yikes, that's no good. Those do field-by-field comparison rather than just ID comparison like we'd want here.

Looking at the code a bit today, it seems we do all this conversion just to build a final product instance. To that end, we can approach this in two ways: one is to kinda shim that together as done here, or we can do some more foundational changes to actually clean up the build process so it's not so abstracted that we lose the goal through the weeds of abstraction calls.
In the interest of getting this through for 4.4.1 and 4.4.2, we can do a bit more of the former, but still kinda lean into what we should do for a more thorough fix in main:

In the `createAnonPayloadAndSignature` method, we can rip out most of its guts and build a map of content IDs to `ProductContent` instances. Using ProductContent gives us a nice wrapper that holds the content and the enabled state already, so we can then use a stable ID as the key to avoid any potential issues there. Streams make this very easy, but we'll have to preload some conversions to make the final code not look like spaghetti:

```
    private String createAnonPayloadAndSignature(Collection<ProductInfo> prodInfo) throws IOException {
        Function<ProductContentInfo, ProductContent> converter = (pcinfo) -> {
            ...
        };

        Function<ProductContent, String> cidFetcher = (pcinfo) -> pcinfo.getContent().getId();

        Map<String, ProductContent> activeContent = prodInfo.stream()
            .filter(Objects::nonNull)
            .flatMap(ProductInfo::getProductContent)
            .filter(Objects::nonNull)
            .map(converter)
            .collect(Collectors.toMap(cidFetcher, Function.identity(), (v1, v2) -> {
                return new ProductContent()
                    .setContent(v2.getContent()) // last in wins
                    .setEnabled(v1.isEnabled() || v2.isEnabled());
            }));

        PromotedContent promotedContent = new PromotedContent(new AnonymousContentPrefix());

        byte[] data = this.createContentAccessDataPayload(null, activeContent, promotedContent);
        return this.createPayloadAndSignature(data);
    }
```

From there, now we just need to update `createContentAccessDataPayload` to accept it, which is actually less work than what we're doing now:

```
    private byte[] createContentAccessDataPayload(Consumer consumer, 
        Map<String, ProductContent> activeContent, PromotedContent promotedContent) {

        Product engProduct = new Product()
            .setId("content_access_eng")
            .setName("Content Access Eng")
            .setProductContent(activeContent.values());

        Product skuProduct = new Product()
            ...
```

We'll also need to update the two other places this is called to also provide the same map:

```
        Function<Map.Entry<Content, Boolean>, ProductContent> pcbuilder = (entry) -> new ProductContent()
          .setContent(entry.getKey())
          .setEnabled(entry.getValue());
          
        Function<ProductContent, String> cidFetcher = (pcinfo) -> pcinfo.getContent().getId();

        Map<String, ProductContent> ownerContent = this.ownerContentCurator
            .getActiveContentByOwner(owner.getId())
            .keySet()
            .stream()
            .map(pcbuilder)
            .collect(Collectors.toMap(cidFetcher, Function.identity(), (v1, v2) -> {
                return new ProductContent()
                    .setContent(v2.getContent()) // last in wins
                    .setEnabled(v1.isEnabled() || v2.isEnabled());
            }));

        byte[] payloadBytes = createContentAccessDataPayload(consumer, ownerContent, promotedContent);  
```

Realistically, that `getActiveContentByOwner` function should be rewritten to kick out ProductContent instances and use a proper CTE to avoid recursive queries locally (since we're already using a native query anyway), but that's a bit of a heavy lift for this branch and 4.4.2.

Reply:
Updated